src/core/wm.cpp

```diff
@@ -109,14 +111,26 @@ void wayfire_focus::init()
     };
     output->connect_signal("wm-focus-request", &on_wm_focus_request);
 
-    on_button = [=] (const wf::buttonbinding_t&)
+    on_button.set_callback([=] (wf::signal_data_t *data)
     {
-        this->check_focus_surface(wf::get_core().get_cursor_focus());
+        auto ev = static_cast<
+            wf::input_event_signal<wlr_event_pointer_button>*>(data);
 
-        return false;
-    };
-    output->add_button(
-        wf::create_option_string<wf::buttonbinding_t>("BTN_LEFT"), &on_button);
+        if (ev->event->state != WLR_BUTTON_PRESSED)
+        {
+            return;
+        }
+
+        if ((!focus_modifiers && wf::get_core().get_keyboard_modifiers()) ||
+            ((ev->event->button == BTN_MIDDLE) && !focus_btn_middle) ||
+            ((ev->event->button == BTN_RIGHT) && !focus_btn_right))
+        {
+            return;
+        }
+
```

Comment:
There are other buttons, for example BTN_EXTRA, BTN_FORWARD, etc.
I don't think we should change focus with such buttons.

Also, what do you think of creating an activator binding, which is used to match the button? I.e the user would list all the buttons they want to cause focus change, and then this focus plugin here will just match the button against the activator.

Reply:
Yes, that's a good idea; it actually makes the result more flexible. I've made this into an option. There are two small UI issues with this though:
 - any modifiers set are ignored (although I don't know why anyone would want to set them; but this could be easily added)
 - when changing this option in wcm, there is a scary warning message about setting a mouse button binding without any modifiers